SI-8976 MutableList.tail.iterator.size is length

som-snytt opened the pull request:
This is fixed by capping the length of iterator.

That is OK because mutating the list by adding to it during
iteration is not allowed.

A test is added that will no doubt be superseded by the QCC tests.
(Quasi-Comprehensive Collections.)

Review by @Ichoran with screws tightened by @retronym

som-snytt committed:
SI-8976 MutableList.tail.iterator.size is length

The previous behavior was to iterate over the mutated
list of arbitrary length. The previous iteration of
the iterator would also iterate the terminal element
of the list without halting.

This is fixed by capping the length of iterator.

That is OK because mutating the list by adding to it during
iteration is not recommended.

For good measure, the exhausted iterator does not hold
a reference to any remaining tail.

A test is added that will no doubt be superseded by the QCC tests.
(Quasi-Comprehensive Collections.)

The test just checks that the extra tail is not strongly
reachable from the iterator. If the garbage collector happens
to kick in and determine that the object is weakly reachable,
then the check terminates early.

som-snytt commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 63196414)
:cat: Roger! Rebuilding pr-scala for 6b7d8e04. :rotating_light:

Ichoran commented:
I don't think this is the right way to fix the problem.  It's certainly simple, which is good, but it's an obligatory two object creations: first create the iterator, then create the wrapper to take at most N.  This also results in everything going through forwarders.  I haven't profiled it, but this is the sort of thing that almost always causes noticable performance regressions.

`LinkedList` exposes its guts to the world (which is why it's deprecated).  This should instead reimplement `LinkedList#iterator` in a way that's limited by length.  This way you'll get only one layer of wrapping and object creation instead of two.

som-snytt commented:
The less simple iteration has the other virtue of not holding onto the remaining tail when it is done, and also of throwing when overtaxed.

som-snytt commented:
Repushed for Java 6 compatibility.  I can't wait for Scala 2.12!  Java 8 support will be nice, but losing Java 6 will be even better.

som-snytt commented:
PLS REBUILD 7029bb7

scala-jenkins commented:
(kitty-note-to-self: ignore 65335799)
:cat: Roger! Rebuilding  for bbdecb29, 9913cffe, 3f13aa8f, 7029bb7d. :rotating_light:

som-snytt commented:
Rebased and added test for the new asserts, and disabled the scala.ref version, which doesn't work.

I had previously noticed that a downside of scala.ref is that it's easy to accidentally create a strong ref.

Removed the unrelated test that seems to be hanging.

som-snytt commented:
ping @Ichoran This won't change anybody's life, but assertCollectable is handy -- once I can use it in other tests.

som-snytt commented:
I wonder what's going on here with jenkins, @adriaanm ?  Is there a wizard who might have a special insight?  Should I check behind the curtain?

adriaanm commented:
PLS SYNCH behind the curtain

scala-jenkins commented:
(kitty-note-to-self: ignore 65540639)
:cat: Synchronaising! :pray:

som-snytt commented:
Since @Ichoran has seasonal employment as a mall Santa, I wonder if @retronym is interested in the `assertCollectable` test helper?

Ichoran commented:
@som-snytt Sorry, both not feeling well and have another important deadline on the 15th; I am not in a position to give this the attention it deserves right now.

som-snytt commented:
@Ichoran I prefer my excuse for you, but do feel better.

FTR, I don't want to sneak in the test method and then start breaking the build because my assumptions about garbage collection were wrong. (As happens sometimes.)

Ichoran commented:
@som-snytt You can monitor GC cycles with `java.lang.management.GarbageCollectorMXBeans` (of which you can get one for full GCs from`jlm.ManagementFactory.getGarbageCollectorMXBeans`).  That might be a gentler way to check your assumptions about garbage collection or wait for a certain number of GC cycles to actually happen.

som-snytt commented:
@Ichoran the test rig just waits on the reference queue, trying to avoid assumptions about when refs are cleared, in relation to gc, finalize, etc.  But it may be harmless to ensure we wait until at least one gc.  Or two?  Well, a long timeout on the ref queue was meant to obviate those questions. But what if Jenkins decides to run with a huge heap and a gc cycle is unimaginably long? OK, it might as well check for one measly gc before failing. Now go get some soup.

retronym commented:
How does OpenJDK write this sort of test? (Do they?)

I wonder if we ought to do this with a Java Agent, which might let us ask the VM directly for the reference count of an object.

retronym commented:
This is the sort of API I was looking for: http://docs.oracle.com/javase/6/docs/platform/jvmti/jvmti.html#FollowReferences

som-snytt commented:
Yes, I was just looking at the heap functions. I also just signed up for the jdk9-dev list. In for a penny.

My only question so far was whether the gc function is forced in a way that System.gc is not. Besides that, I don't immediately see anything that yields more info before reachability is determined at gc. I'll pursue it later, though.

retronym commented:
A less scarifying possibility is to use Java reflection to find the transitive closure of referenced objects.

``` scala
scala> def refs(o: AnyRef): List[AnyRef] = {
     |   val seen = new java.util.IdentityHashMap[AnyRef, Unit]()
     |   def loop(o: AnyRef) {
     |     println(o)
     |     if (!seen.containsKey(o)) {
     |       seen.put(o, ())
     |       for {
     |         f <- o.getClass.getDeclaredFields
     |         if !Modifier.isStatic(f.getModifiers)
     |         if !f.getType.isPrimitive
     |         val contents = {f.setAccessible(true); f.get(o)}
     |       } loop(contents)
     |     }
     |   }
     |   loop(o)
     |   import scala.collection.JavaConverters._
     |   seen.keySet.asScala.toList
     | }
warning: there was one deprecation warning; re-run with -deprecation for details
refs: (o: AnyRef)List[AnyRef]

scala> val l = List(1, 2, 3, 4)
l: List[Int] = List(1, 2, 3, 4)

scala> val it = l.iterator
it: Iterator[Int] = non-empty iterator

scala> it.next()
res28: Int = 1

scala> refs(it)
non-empty iterator
List(2, 3, 4)
2
List(3, 4)
3
List(4)
4
List()
List(1, 2, 3, 4)
1
List(2, 3, 4)
res29: List[AnyRef] = List(non-empty iterator, 2, 3, List(), List(3, 4), List(2, 3, 4), 1, List(4), List(1, 2, 3, 4), 4)
```

som-snytt commented:
That's good if you know where to start looking.

For instance, my scala.ref test failed because my object ref was strongly captured somewhere.  I'm sure it's pretty easy to have intermediate structures on the stack or a closure. Maybe I'll try to understand that syndrome better.

Following refs can't yield false negatives, anyway.

som-snytt commented:
@retronym a test that just follows strong refs. For fun, it also checks a weak ref.

lrytz commented:
Related problem, which persists after this change:

``` scala
scala> import scala.collection._

scala> val l = mutable.MutableList(1,2)
l: scala.collection.mutable.MutableList[Int] = MutableList(1, 2)

scala> val t = l.tail
t: scala.collection.mutable.MutableList[Int] = MutableList(2)

scala> l += 3
res0: l.type = MutableList(1, 2, 3)

scala> l.toList
res2: List[Int] = List(1, 2, 3)

scala> t.toString
res3: String = MutableList(2)

scala> t.toList
res4: List[Int] = List(2, 3)
```

What is the contract of tail? Is it supposed to return a new mutable list, or one that shares data structures with the original one? Because if elements are changed, this is reflected in both. Why should it not be the case for the number of elements when added?

``` scala
scala> val l = mutable.MutableList(1,2)
l: scala.collection.mutable.MutableList[Int] = MutableList(1, 2)

scala> val t = l.tail
t: scala.collection.mutable.MutableList[Int] = MutableList(2)

scala> t(0) = 10

scala> l
res6: scala.collection.mutable.MutableList[Int] = MutableList(1, 10)

scala> t
res7: scala.collection.mutable.MutableList[Int] = MutableList(10)
```

Given that and also @Ichoran 's performance concerns, I'll push this to 2.11.6

som-snytt commented:
@lrytz The quick original solution @Ichoran objected to was just to return `first.iterator take length`. This formulation is what he asked for.

The desirable behavior is for `list.length == iterator.length`. The list knows how long it is.

lrytz commented:
OK, thanks for the clarification. But the `toList` behavior is still an issue, no?

som-snytt commented:
I'm sure there are issues. Appending using the tail alias breaks appending. So the lesson is, caveat programmator. But they already know that.

```
scala> val a = mutable.MutableList(1,2,3)
a: scala.collection.mutable.MutableList[Int] = MutableList(1, 2, 3)

scala> a += 4
res7: a.type = MutableList(1, 2, 3, 4)

scala> a
res8: scala.collection.mutable.MutableList[Int] = MutableList(1, 2, 3, 4)

scala> a ++= List(8,9,10)
res9: a.type = MutableList(1, 2, 3, 4, 8, 9, 10)

scala> val b = a.tail
b: scala.collection.mutable.MutableList[Int] = MutableList(2, 3, 4, 8, 9, 10)

scala> b.length
res10: Int = 6

scala> a.length
res11: Int = 7

scala> a ++= List(8,9,10)
res12: a.type = MutableList(1, 2, 3, 4, 8, 9, 10, 8, 9, 10)

scala> b += 7
res13: b.type = MutableList(2, 3, 4, 8, 9, 10, 7)

scala> a
res14: scala.collection.mutable.MutableList[Int] = MutableList(1, 2, 3, 4, 8, 9, 10, 7)

scala> b
res15: scala.collection.mutable.MutableList[Int] = MutableList(2, 3, 4, 8, 9, 10, 7)

scala> a ++= List(8,9,10)
res16: a.type = MutableList(1, 2, 3, 4, 8, 9, 10, 7)
```

som-snytt commented:
BTW, for eager operations, doing `b.toList take b.length` is OK. So that much is easy to work around.

Ichoran commented:
LGTM
